core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -516,61 +517,131 @@ class KafkaApis(val requestChannel: RequestChannel,
   def handleOffsetRequest(request: RequestChannel.Request) {
     val correlationId = request.header.correlationId
     val clientId = request.header.clientId
+    val version = request.header.apiVersion()
     val offsetRequest = request.body.asInstanceOf[ListOffsetRequest]
 
-    val (authorizedRequestInfo, unauthorizedRequestInfo) = offsetRequest.offsetData.asScala.partition {
-      case (topicPartition, _) => authorize(request.session, Describe, new Resource(auth.Topic, topicPartition.topic))
-    }
+    val mergedResponseMap = {
```

Comment:
Actually, there's quite a bit of redundance in the handling of the two versions. For example, the authorization checks, the leader checking, and some of the exception handling look the same. Seems like we ought to be able to do better.